packages/cozy-harvest-lib/src/services/budget-insight.js

```diff
@@ -244,35 +244,49 @@ export const resumeBIConnection = flow => {
 }
 
 /**
- * Checks for 2FA or decoupled requests
+ * @typedef biConnection
+ * @property {number} id
+ * @property {number} id_user
+ * @property {number} id_connector
+ * @property {string|null} state  - ( wrongpass | additionalInformationNeeded | websiteUnavailable | actionNeeded | SCARequired | decoupled | passwordExpired | webauthRequired | rateLimiting | bug). Null indicates a success
+ * @property {string|null} last_update - Date string: Last successful update.
+ * @property {string|null} created - Date string: Creation date
+ * @property {boolean} active - Whether this connection is active and will be automatically synced.
+ * @property {string|null} last_push - Date string: Last successfull push
+ * @property {string|null} next_try - Date string: Date of next synchronization.
+ */
+
+/**
+ * Checks for any number 2FA and/or decoupled requests
  *
  * - Resolves when 2FA has been resolved
  * - Triggers LOGIN_SUCCESS_EVENT
+ *
+ * @param  {Object} options
+ * @param  {biConnection} options.biConnection
+ * @param  {ConnectionFlow} options.flow
  */
-export const finishConnection = async ({ biConnection, flow }) => {
-  if (biConnection.error) {
-    if (biConnection.error === DECOUPLED_ERROR) {
+export const finishConnection = async ({ flow }) => {
+  while (flow.getData().biConnection.error) {
```

Comment:
I don't know the code too much to have a strong opinion, but with this kind of loop we can get stuck for a long time? Can we have some kind of delay, for instance, in order to throw an error after X minutes ?

Reply:
Since the in this case, the loop is asynchronous and not connector is launched yet. The user still has control on the navigation.
I don't think this is necessary here.
I wondered if i should add a limit on the number of loops but couldn't define a proper limit and proper associated error message.